### PR TITLE
build: response to comments and spec references

### PR DESCRIPTION
**Pull request by nicks**
Hello @maiamcc,

Please review the following commits I made in branch nicks/asserts2:

b3e9c6f6776e70fe312dea72d36b320fe192f65c (2018-08-15 18:20:10 -0400)
build: response to comments and spec references

### CODE REVIEW

**Comment by nicks** [comment]
@maiamcc i think we need to walk thru the failing test tomorrow. I'm pretty sure this test *should* be failing, but i don't totally understand what you were trying to test here.

**Comment by nicks** [comment]
(in particular, the problem test is TestEntrypoint
https://github.com/windmilleng/tilt/blob/master/internal/build/docker_test.go#L269
which is currently asserting that BuildDocker runs the entry point, which seems wrong)

File: internal/build/docker.go [hunk 1]
```diff
@@ -229,12 +229,23 @@ func (l *localDockerBuilder) imageWithEntrypoint(ctx context.Context, d digest.D
 // startContainer starts a container from the given image ref, exec'ing a command if given.
 // Returns the container id iff the container successfully starts up; else, error.
 func (l *localDockerBuilder) startContainer(ctx context.Context, imgRef string, cmd *tiltd.Cmd) (cId string, err error) {
+
 	config := &container.Config{Image: imgRef}
+
+	// In Docker, both the Entrypoint and the Cmd are used to determine what
+	// process the container runtime uses, where Entrypoint takes precedence over
+	// container. We set both here to ensure that we don't get weird results due
```
Comment (maiamcc) [inline internal/build/docker.go]: believe you meant "entrypoint takes precedence over cmd"

**Comment by nicks** [comment]
per our offline discussion, we realized that there's a problem where the same interface is trying to use the same API for two different use-cases: run a container without an entrypoint, and run a container with the pre-existing entrypoint. I rejiggered the API to support both.

File: internal/build/container.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+	// In Docker, both the Entrypoint and the Cmd are used to determine what
+	// process the container runtime uses, where Entrypoint takes precedence over
+	// command. We set both here to ensure that we don't get weird results due
+	// to inheritance.
+	//
+	// If cmd is nil, we use a fake cmd that does nothing.
+	//
+	// https://github.com/opencontainers/image-spec/blob/master/config.md#properties
+	if len(cmd.Argv) > 0 {
```
Comment (maiamcc) [inline internal/build/container.go]: we have a `cmd.Empty()` func now fwiw

File: internal/build/docker.go [hunk 3]
```diff
@@ -239,16 +239,9 @@ func (l *localDockerBuilder) imageWithEntrypoint(ctx context.Context, img digest
...
-func (l *localDockerBuilder) startContainer(ctx context.Context, imgRef string, cmd *model.Cmd) (cId string, err error) {
-	config := &container.Config{Image: imgRef}
-	if cmd != nil {
-		config.Cmd = cmd.Argv
-	} else {
-		config.Cmd = []string{"sh", "-c", "# NOTE(nick): a fake entrypoint"}
-	}
-
+// startContainer starts a container from the given config.
+// Returns the container id iff the container successfully starts up; else, error.
```
Comment (maiamcc) [inline internal/build/docker.go]: s/successfully starts up/successfully runs

File: internal/build/container.go (hunk 2 above)
Comment (nicks) [reply internal/build/container.go]: done!

File: internal/build/docker.go (hunk 3 above)
Comment (nicks) [reply internal/build/docker.go]: done

**Pull request merged**